pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+	app, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	appID := app.Name
+	err := appMgr.amProtocol.RemoveApplication(appID)
+	if err != nil {
+		log.Logger.Error("Application removal failed",
+			zap.String("appID", appID),
+			zap.Error(err))
+	}
+	log.Logger.Debug("App CRD deleted",
+		zap.String("Name", appID))
+	//TODO: delete related pods and deployments
+}
+
+/*
+Add application to scheduler
+*/
+func (appMgr *AppManager) addApp(obj interface{}) {
+	appCRD, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	if appMeta, ok := appMgr.getAppMetadata(appCRD); ok {
+		app := appMgr.amProtocol.GetApplication(appMeta.ApplicationID)
+		if app == nil {
+			appMgr.amProtocol.AddApplication(&interfaces.AddApplicationRequest{
+				Metadata: appMeta,
+				//TODO: debug recovery
+				//TODO: try to define the owner reference for the submitted pods to this CRD and try of we delete the
+				//CRD the pods will be deleted as well
+				Recovery: false,
+			})
+			// set and save status = New
+			appMgr.updateAppCRDStatus(appCRD, appv1.NewApplicationState)
+		}
+	}
+}
+
+func (appMgr *AppManager) updateAppCRDStatus(appCRD *appv1.Application, status appv1.ApplicationStateType) {
+	copy := appCRD.DeepCopy()
+	copy.Status = appv1.ApplicationStatus{
+		AppStatus:  status,
+		Message:    "app CRD status change",
+		LastUpdate: v1.NewTime(time.Now()),
+	}
+	_, err := appMgr.controller.appClient.ApacheV1alpha1().Applications(appCRD.Namespace).UpdateStatus(copy)
+	if err != nil {
+		log.Logger.Error("Failed to update application CRD",
+			zap.String("AppId", copy.Name))
+		return
+	}
+}
+
+func (appMgr *AppManager) getAppMetadata(app *appv1.Application) (interfaces.ApplicationMetadata, bool) {
+	appID := app.Name
+
+	// tags will at least have namespace info
+	// labels or annotations from the pod can be added when needed
+	// user info is retrieved via service account
+	tags := map[string]string{}
+	if app.Namespace == "" {
+		tags[constants.AppTagNamespace] = constants.DefaultAppNamespace
+	} else {
+		tags[constants.AppTagNamespace] = app.Namespace
+	}
+
+	return interfaces.ApplicationMetadata{
+		ApplicationID: appID,
+		QueueName:     app.Spec.Queue,
+		User:          "default",
+		Tags:          tags,
+	}, true
+}
+
+func convertShimAppStateToAppCRDState(status string) appv1.ApplicationStateType {
+	switch status {
+	case corecache.New.String():
+		return appv1.NewApplicationState
+	case corecache.Accepted.String():
+		return appv1.AcceptedState
+	case corecache.Starting.String():
+		return appv1.StartingState
+	case corecache.Running.String():
+		return appv1.RunningState
+	case corecache.Waiting.String():
+		return appv1.WaitingState
+	case corecache.Rejected.String():
+		return appv1.RejectedState
+	case corecache.Completed.String():
+		return appv1.CompletedState
+	case corecache.Killed.String():
+		return appv1.KilledState
+	default:
+		return "Undefined"
+	}
```

Comment:
I think this is one of the main things we need to fix.
We need to make shim side app state in sync with the core side states, pls take a look at the core side status: http://yunikorn.apache.org/docs/next/design/scheduler_object_states#application-state.
I would even suggest creating a separate JIRA to get this fixed first.
Ideally, all the state transition are done in the core side, and core leverages UpdateApplication API to send the state updates to the shim, and that triggers the transition of shim side app-state-machine.

Reply:
I completely agree with you that we should keep the states in synch. For me this one seems a bigger refactoring task. And also we have different statuses for the shim side and different ones for the core side:
https://github.com/apache/incubator-yunikorn-k8shim/blob/master/pkg/cache/application.go#L71-L101
https://github.com/apache/incubator-yunikorn-core/blob/master/pkg/cache/application_state.go#L70-L100
If we want refactor it now, after then it will be easier to handle the AppCRD status as well, than it is now.

Reply:
Based on the offline discussion, we will not do the refactoring for now given the complexity of this problem.